src/coreclr/vm/eventtrace.cpp

```diff
@@ -6976,20 +6993,48 @@ VOID ETW::MethodLog::SendMethodRichDebugInfo(MethodDesc* pMethodDesc, PCODE pNat
     ULONG32 numMappings = 0;
     if (DebugInfoManager::GetRichDebugInfo(request, fpNew, NULL, &inlineTree, &numInlineTree, &mappings, &numMappings))
     {
-        unsigned dataSize = 8 + numInlineTree * sizeof(ICorDebugInfo::InlineTreeNode) + numMappings * sizeof(ICorDebugInfo::RichOffsetMapping);
+        static_assert_no_msg((std::is_same<decltype(inlineTree->Method), CORINFO_METHOD_HANDLE>::value));
+        static_assert_no_msg((std::is_same<decltype(inlineTree->ILOffset), uint32_t>::value));
+        static_assert_no_msg((std::is_same<decltype(inlineTree->Child), uint32_t>::value));
+        static_assert_no_msg((std::is_same<decltype(inlineTree->Sibling), uint32_t>::value));
+
+        static_assert_no_msg((std::is_same<decltype(mappings->ILOffset), uint32_t>::value));
+        static_assert_no_msg((std::is_same<decltype(mappings->Inlinee), uint32_t>::value));
+        static_assert_no_msg((std::is_same<decltype(mappings->NativeOffset), uint32_t>::value));
+
+        const uint32_t inlineTreeNodeDataSize =
+            sizeof(CORINFO_METHOD_HANDLE) +
+            sizeof(uint32_t) * 3;
+
+        const uint32_t richOffsetMappingDataSize =
+            sizeof(uint32_t) * 3 +
+            1; // source type
+
+        unsigned dataSize = 8 + numInlineTree * inlineTreeNodeDataSize + numMappings * richOffsetMappingDataSize;
 
         InlineSBuffer<1024> buffer;
         buffer.Preallocate(dataSize);
 
         BYTE* pBuffer = &buffer[0];
-        memcpy(pBuffer, &numInlineTree, 4);
-        pBuffer += 4;
-        memcpy(pBuffer, &numMappings, 4);
-        pBuffer += 4;
-        memcpy(pBuffer, inlineTree, numInlineTree * sizeof(ICorDebugInfo::InlineTreeNode));
-        pBuffer += numInlineTree * sizeof(ICorDebugInfo::InlineTreeNode);
-        memcpy(pBuffer, mappings, numMappings * sizeof(ICorDebugInfo::RichOffsetMapping));
-        pBuffer += numMappings * sizeof(ICorDebugInfo::RichOffsetMapping);
+        WriteToBuffer(&pBuffer, numInlineTree);
+        WriteToBuffer(&pBuffer, numMappings);
+        for (uint32_t i = 0; i < numInlineTree; i++)
+        {
+            WriteToBuffer(&pBuffer, inlineTree[i].Method);
```

Comment:
Use of WriteToBuffer is a concern for me.

1. The definition of these types does not have comments/description that the field's types are part of a contract that is never permitted to change.
2. This is an inefficient encoding. For instance, ILOffset is unlikely to use the full 32 bits that is being sent here, and the Child/Sibling are highly unlikely to be large as well.

Reply:
We discussed offline. I'm ok with the current set of changes.